### PR TITLE
refactored modules

### PR DESCRIPTION
**Pull request by displague**
The purpose is to refactor the TF config into separate master, node, and instance modules so that the basic/common, and more opinionated, K8s provisioning can be reused outside of this project.  Another benefit would be more parallelization of instance provisioning.

TODO:
- [x] add descriptions to all variables
- [x] verify all variables are making it through to their ancestor modules
- [x] add a readme for each sub-module
- [x] update the base readme to describe the modules
- [x] verify master and nodes provision in parallel

indentation and field order changes are due to `terrraform fmt`

### CODE REVIEW

**Comment by displague** [comment]
I'd really like to have the example in this repo reflect use of the Helm Terraform provider, but I am currently running into https://github.com/terraform-providers/terraform-provider-helm/issues/77 (or something that presents as that).

**Comment by displague** [comment]
The external data [`kubeadm_join`](https://github.com/linode/terraform-linode-k8s/pull/23/files#diff-b1d016ca6da40e76e6cb94a2509cbadcR53) and its output variable [`kubeadm_join_command`](https://github.com/linode/terraform-linode-k8s/pull/23/files#diff-49179c3c2ad9a79e8fb9a79b612c45fdR11) are running into this problem on destroy:

```
Error: Error applying plan:

1 error(s) occurred:

* module.masters.output.kubeadm_join_command: Resource 'data.external.kubeadm_join' does not have attribute 'result' for variable 'data.external.kubeadm_join.result'

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

$ terraform destroy -force
data.linode_instance_type.type: Refreshing state...
data.linode_instance_type.type: Refreshing state...
linode_instance.instance: Refreshing state... (ID: 12451578)

Error: Error applying plan:

1 error(s) occurred:

* module.masters.output.kubeadm_join_command: variable "kubeadm_join" is nil, but no error was reported

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

This appears to be https://github.com/hashicorp/terraform/issues/17862 (and related issues).

```sh
$ TF_WARN_OUTPUT_ERRORS=0 terraform destroy -force
data.linode_instance_type.type: Refreshing state...
data.linode_instance_type.type: Refreshing state...
linode_instance.instance: Refreshing state... (ID: 12451578)
module.masters.module.master_instance.linode_instance.instance: Destroying... (ID: 12451578)
module.masters.master_instance.linode_instance.instance: Still destroying... (ID: 12451578, 10s elapsed)
module.masters.master_instance.linode_instance.instance: Still destroying... (ID: 12451578, 20s elapsed)
module.masters.master_instance.linode_instance.instance: Still destroying... (ID: 12451578, 30s elapsed)
module.masters.master_instance.linode_instance.instance: Still destroying... (ID: 12451578, 40s elapsed)
module.masters.module.master_instance.linode_instance.instance: Destruction complete after 42s

Destroy complete! Resources: 1 destroyed.
```

**Review by jfrederickson — COMMENTED** [review]
I've been going through this off and on for a bit. It's quite a large diff, but I _think_ everything looks okay. I was able to spin up a cluster locally from this branch, so there's that.

That said, as all of the resources being deployed have changed, re-running `terraform apply` after this change will destroy and recreate the entire cluster. That's kind of Terraform's M.O., but... I wonder if there's a way to prevent this for clusters deployed with a previous version of the provider. (Is this worth worrying about? Should we expect users of this module to lock to a specific version? We should probably update the example in the README to do that if so.)

File: example/main.tf [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+// 
+// ## TODO
+// - Fix timeouts (set better "depends_on" values)
+// - Fix permission issue with helm listing configmaps from the kube-system namespace:
+//   https://github.com/terraform-providers/terraform-provider-helm/issues/77
+
+module "linode_k8s" {
+  # source = "linode/k8s/linode"
+  # version      = "0.0.6"
+  source = "git::https://github.com/displague/terraform-linode-k8s?ref=separate_modules"
```
Comment (asauber) [inline example/main.tf]: this should ref master

**Review by asauber — APPROVED** [review]
other than the ref change, this LGTM!

**Comment by asauber** [comment]
@jfrederickson I think that the performance and flexibility enhancement of this change outweighs the inconvenience of needing to re-provision. That said, we should provide a warning of some sort to existing users.

**Comment by displague** [comment]
Addressed both concerns -- thanks @asauber and @jfrederickson

**Comment by displague** [comment]
@jfrederickson this is way out of date, but `terraform state mv` provides the means to transition from the old resource names to the new ones.

**Comment by displague** [comment]
This work (and the changes that followed) seem to lend themselves towards a Linode example in https://github.com/hashicorp/cluster-api-provider-terraform-cloud/tree/main/examples/.  That example could leverage these modules rather than duplicating infra provisioning in the CAPTC provider.

(An LKE example could also be added, but that would be a more straightforward copy of the GKE pattern)

**Pull request merged**